napi.h

```diff
@@ -1760,6 +1761,32 @@ namespace Napi {
     std::string _error;
   };
 
+  class ThreadSafeFunction {
+  public:
+    using Callback = std::function<void(Napi::Env, Napi::Function)>;
+
+    ThreadSafeFunction();
+    ThreadSafeFunction(napi_env env,
```

Comment:
We use the factory method `::New()` for all our types. We should use one here too. In fact, we should have several factory methods which all call the same constructor and which differ from one another in that `resource_name` can be a `const char*`, a `std::string`, or a `napi_value`.